src/utils/sqs_client.ts

```diff
@@ -0,0 +1,44 @@
+import { SQS } from 'aws-sdk';
+
+import { LONG_POLLING_WAIT_TIME_SECONDS, SINGLE_MESSAGE } from '../constants';
+
+/**
+ * SqsClient wraps SQS, making it far easier to unit test SQS and ignore SQS details
+ */
+export class SqsClient {
+    constructor(private readonly _sqs: SQS, private readonly _queueUrl: string) {}
+
+    public async receiveMessageAsync(): Promise<SQS.Message | null> {
+        const response = await this._sqs
+            .receiveMessage({
+                MaxNumberOfMessages: SINGLE_MESSAGE,
+                WaitTimeSeconds: LONG_POLLING_WAIT_TIME_SECONDS,
+                QueueUrl: this._queueUrl,
+            })
+            .promise();
+
+        if (response?.Messages?.length !== 1) {
+            return null;
+        }
+        return response.Messages[0];
+    }
+
+    public async changeMessageVisibilityAsync(receiptHandle: string, visibilityTimeout: number): Promise<void> {
+        await this._sqs
+            .changeMessageVisibility({
+                QueueUrl: this._queueUrl,
+                ReceiptHandle: receiptHandle,
+                VisibilityTimeout: visibilityTimeout,
+            })
+            .promise();
+    }
+
+    public async deleteMessageAsync(receiptHandle: string): Promise<void> {
+        await this._sqs
```

Comment:
Will this explode if there is a problem deleting the message? what do we expect the behavour to look like?

Reply:
the behavior would be:
- Job in DB is already placed in either `success`, `failed`, or some terminal status
- We fail to delete the message from SQS (i.e. this line blows up)
- Worker dies and gets restarted by k8s
- Message becomes visible after its visibility timeout expires (maybe to another worker)
- We consume the message again

There's no way to entirely prevent against this from happening, but one thing we an do is do some retries. For example, using the https://www.npmjs.com/package/async-retry package.

I actually tried adding that package, but had some issues. As this diff was getting pretty big, I think it makes sense to beef up our systems a bit better in a separate commit

We might also want to define the behavior for what we should do if SQS throws an error. My plan is to kill the worker, but only after N number of retries - which I think should be sufficient